### ISSUE TITLE
generation to capnp & mypy

### ISSUE THREAD

**Issue by despiegk** [Jumpscale/go-raml#344]
```python
class CodeGenerator:
    """
    """
    def ramlschemaToCapnp(self,ramlSchemaPath,capnpSchemaPath=""):
        """
        @param ramlSchemaPath schema of 1 or more models
        @return new file, if capnpSchemaPath unspecified then is same path as capnp but extension became .capnp in stead of raml

        only process the schema parts, ignore the other raml parts

        remarks
        - raml 1.0 (https://github.com/raml-org/raml-spec/blob/master/versions/raml-10/raml-10.md/#raml-data-types)
        - capnnp latest

        """
        templatePath=self._path+"/templates/schema.capnp"
        #use templates to do the conversion (pystache?)

    def ramlschemaToMypy(self,ramlSchemaPath,mypyPath=""):
        """
        @param ramlSchemaPath schema of 1 or more models
        @return new file, if capnpSchemaPath unspecified then is same path as capnp but extension became .capnp in stead of raml

        only process the schema parts, ignore the other raml parts

        remarks
        - raml 1.0
        - mypy: http://mypy.readthedocs.io/en/latest/

        generate 2 classes

        class $NameCollection():
            def new(self,bin=None):
                '''
                bin is capnp binary structure
                if None return an empty one
                '''

        class $NameItem():
            def tocapnp(self):
                '''
                return capnp binary structure
                '''
                

        the code generated for python has following features
        - strict typing
        - proper python classes
        - tocapnp method (generate capnp schemas as well, see above and use that schema to do the conversion, use j.data.capnp tools ...): see above

        """
        templatePath=self._path+"/templates/schemaMypy.py"
        #use templates to do the conversion (pystache?)

```

**Comment by iwanbk** [comment]
@serboctor 
for capnp we already support for it but it is still in early stage, not well tested and maintained.

The capnp generator code can be found on this 
https://github.com/Jumpscale/go-raml/blob/40ae1a6de9e3d051e91ae50866199c36a523469d/codegen/capnp/command.go#L10-L26

It then generator capnp schema using this template https://github.com/Jumpscale/go-raml/blob/40ae1a6de9e3d051e91ae50866199c36a523469d/codegen/templates/struct_capnp.tmpl

It however doesn't use latest parser  improvement which already used by Go, Python, and Nim generator. So the capnp generator only handle `types` defined in `types` field of root raml object, never see `types` defined in request body etc..

The improvement is wrapping the parser to make it easier to use.

You can see how Nim generator uses the wrapper at https://github.com/Jumpscale/go-raml/blob/40ae1a6de9e3d051e91ae50866199c36a523469d/codegen/nim/object.go#L66-L117

**Comment by iwanbk** [comment]
for `ramlschemaToMypy` part, you might also want to do it along with #347

**Comment by serboctor** [comment]
> It however doesn't use latest parser improvement which already used by Go, Python, and Nim generator. So the capnp generator only handle types defined in types field of root raml object, never see types defined in request body etc..

I think if the capnp command is used for strictly generating schemas, then it should only handle the types defined in the `types` field, it is after all not generating some client/server code, it is simply converting standalone raml types to capnp schemas. Also, in the specs it says `only process the schema parts, ignore the other raml parts` so I assume this implies only process the types in the `types` field, @despiegk  did I get it right?

>for ramlschemaToMypy part, you might also want to do it along with #347

sure thing.

**Comment by serboctor** [comment]
@despiegk for the mypy generation, we can use the existing python classes generated by the go-raml python client generation and extend them to add mypy type in the variable assignments, and the `tocapnp` function. It will look like the following generated code where `set_proprty` function is a helper function we use to validate that the data are of expected types and raises valueErrors during run time.
```"""
Auto-generated class for Cluster
"""
from .EnumClusterDriveType import EnumClusterDriveType
from .EnumClusterStatus import EnumClusterStatus
from .StorageServer import StorageServer

from . import client_support


class Cluster(object):
    """
    auto-generated. don't touch.
    """

    @staticmethod
    def create(**kwargs):
        """
        :type clusterType: StorageServer
        :type driveType: EnumClusterDriveType
        :type label: int, str
        :type nodes: list[str]
        :type status: EnumClusterStatus
        :type storageServers: dict
        :rtype: Cluster
        """

        return Cluster(**kwargs)

    def __init__(self, json=None, **kwargs):
        if json is None and not kwargs:
            raise ValueError('No data or kwargs present')

        class_name = 'Cluster'
        data = json or kwargs
        data_types = [StorageServer]
        self.clusterType = client_support.set_property('clusterType', data, data_types, False, [], False, True, class_name) # type: Union[data_types]
        data_types = [EnumClusterDriveType]
        self.driveType = client_support.set_property('driveType', data, data_types, False, [], False, True, class_name) # type: Union[data_types]
        data_types = [int, str]
        self.label = client_support.set_property('label', data, data_types, False, [], False, True, class_name) # type: Union[data_types]
        data_types = [str]
        self.nodes = client_support.set_property('nodes', data, data_types, False, [], True, True, class_name) # type: Union[data_types]
        data_types = [EnumClusterStatus]
        self.status = client_support.set_property('status', data, data_types, False, [], False, True, class_name) # type: Union[data_types]
        data_types = [dict]
        self.storageServers = client_support.set_property('storageServers', data, data_types, False, [], False, True, class_name) # type: Union[data_types]

    def __str__(self):
        return self.as_json(indent=4)

    def as_json(self, indent=0):
        return client_support.to_json(self, indent=indent)

    def as_dict(self):
        return client_support.to_dict(self)

    def tocapnp(self):
        """
            return capnp binary structure
        """
        # do some capnp magic here
```

Does this work with what you had in mind?

**Comment by serboctor** [comment]
Since mypy doesnt offer dynamic checking, it is not of much use to us. I suggested to @despiegk adding the tocapnp functions to the regular python classes we generate in go-raml. I also introduced `python-capnp` command in case someone wants to generate those classes without having to generate a whole server/client. These changes are part of this pull request https://github.com/Jumpscale/go-raml/pull/380
